bip-psbt.mediawiki

```diff
@@ -0,0 +1,245 @@
+!Field Size
+!Name
+!Type
+!Value
+!Description
+|-
+| 4
+| Magic Bytes
+| int32_t
+| <tt>0x70736274</tt>
+| Magic bytes which are ASCII for psbt <ref>'''Why use 4 bytes for psbt?''' The
+transaction format needed to start with a 5 byte header which uniquely identifies
+it. The first bytes were chosen to be the ASCII for psbt because that stands for
+Partially Signed Bitcoin Transaction. </ref>
+|-
+| 1
+| separator
+| char
+| <tt>0xff</tt>
+| Must be <tt>0xff</tt> <ref>'''Why Use a separator after the magic bytes?''' The separator
+is part of the 5 byte header for PSBT. This byte is a separator of <tt>0xff</tt> because
+this will cause any non-PSBT unserializer to fail to properly unserialize the PSBT
+as a normal transaction. Likewise, since the 5 byte header is fixed, no transaction
+in the non-PSBT format will be able to be unserialized by a PSBT unserializer.</ref>
+|-
+| 1+
+| Transaction
+| key-value pair
+| varies
+| A transaction type key-value pair. This field must exist.
+|-
+| 1+
+| Redeem Scripts
+| Array of key-value pairs
+| varies
+| All of the redeem script type key-value pairs
+|-
+| 1+
+| Witness Scripts
+| Array of key-value pairs
+| varies
+| All of the witness script type key-value pairs
+|-
+| 1
+| separator
+| char
+| <tt>0x00</tt>
+| Must be <tt>0x00</tt> <ref>'''Why is the separator here <tt>0x00</>> instead of <tt>0xff</tt>?'''
+The separator here is used to distinguish between each chunk of data. A separator of 0x00 would mean that
+the unserializer can read it as a key length of 0, which would never occur with actual keys. It can thus
+be used as a separator and allow for easier unserializer implementation.</ref>
+|-
+| 1+
+| Inputs
+| Array of inputs
+| varies
+| The key-value pairs for each input as described below
+|}
+
+Each Input item is specified as follows:
+
+{| class="wikitable" style="width: auto; text-align: center; font-size: smaller; table-layout: fixed;"
+!Field Size
+!Name
+!Type
+!Value
+!Description
+|-
+| 1+
+| Non-Witness UTXO
+| key-value pair
+| varies
+| A Non-Witness UTXO type key-value pair
+|-
+| 1+
+| Witness UTXO
+| key-value pair
+| varies
+| A Witness UTXO type key-value pair
+|-
+| 1+
+| Partial Signatures
+| Array of key-value pairs
+| varies
+| All of the partial signature type key-value pairs for this input
+|-
+| 1+
+| Public Key derivation paths
+| Array of key-value pairs
+| varies
+| All of the BIP 32 Derivation Path type key-value pairs for this input
+|-
+| 1
+| separator
+| char
+| <tt>0x00</tt>
+| Must be <tt>0x00</tt>.
+|}
+
+Each block of data between separators can be viewed as a scope, and all separators
+are required<ref>'''Why are all separators required?''' The separators are required
+so that the unserializer knows which input it is unserializing data for.</ref>.
+Within each scope, the type of each key-value pair must be greater
+than or equal to the pair before it<ref>'''Why must the types increase?''' The types
```

Comment:
If you want to maintain this property that signing can be done by iterating over the fields while deserializing once, it may be necessary to add some gaps between the field numbers. It may be needed to insert new things in between the existing ones?